Add blasons managment in carrousel

a2line opened the pull request:
This PR mainly adds a support for coat of arms in the carrousel. It allows to avoid to duplicate images: A coat of arm can be added from an individual to one of his descendant instead of adding the same coat of arms as portrait to each individuals. The original idea and work was done last year by @mdelambilly (thank you very much again!) It was later reworked and also rebased three times to master (to support image map, base reorganisation and other changes…) by @hgouraud and myself during the year. It should be nearly usable as it is yet but I add it as a temporary draft because it needs far more testings.

Note as reminder that the last work on `notes_album.txt` is not present. It should be added back and merged with `notes_gallery.txt`.

mdelambilly committed:
Add blasons managment in carrousel

a2line commented:
# Code Review
After reviewing the `blason_init.diff` file, I've identified several aspects that need attention. This patch adds coat of arms ("blason") functionality to Geneweb, allowing users to associate heraldic emblems with individuals similar to how portraits work, with the added capability of inheritance through patrilineal lines.

## Overall Structure and Approach

The implementation follows a solid approach by extending the existing image handling system rather than creating parallel systems. The code:

1. Extends the existing image handling code to support blasons
2. Adds UI elements for managing blasons
3. Implements inheritance logic for blasons along patrilineal lines
4. Adds blason display to various templates

## Specific Issues and Recommendations

### 1. Function and Variable Naming Inconsistencies

The codebase mixes terminology between "image," "portrait," and "blason" in ways that can be confusing:

```ocaml
(* Examples of naming inconsistency *)
val default_image_filename : string -> base -> person -> string  (* Good generic name *)
let get_portrait conf base p = get_portrait_aux conf base p false  (* Specific to portraits *)
```

**Recommendation:** Review naming across the codebase to ensure consistent terminology. Consider using "image" for generic functions and "portrait"/"blason" for specific implementations.

### 2. Path Handling Complexities

The path handling logic is complex and potentially fragile:

```ocaml
let path_of_filename conf fname =
  let fname1 = Filename.concat (!GWPARAM.images_d conf.bname) fname in
  if Sys.file_exists fname1 then fname1  (* Return string *)
  else Util.search_in_assets (Filename.concat "images" fname)  (* Return string *)
```

Previously this returned a `Path` variant, but now returns a string. This inconsistency could lead to bugs at call sites.

**Recommendation:** Ensure all code that previously used `Path` variants is updated to work with strings. Consider adding unit tests for path handling.

### 3. File Operation Error Handling

File operations have inconsistent error handling:

```ocaml
let move_file_to_save dir file =
  try
    (* FIXME attention, le basename détruit les sous dossiers *)
    let save_dir = Filename.concat dir "saved" in
    if not (Sys.file_exists save_dir) then File.create_dir ~parent:true save_dir;
    let fname = Filename.basename file in
    let orig_file = Filename.concat dir fname in
    let saved_file = Filename.concat save_dir fname in
    (* TODO handle rn errors *)
    rn orig_file saved_file;
    (* ... *)
    1
  with _ -> 0
```

The error handling is minimal (returning 0 for any exception) and there's a comment saying "TODO handle rn errors".

**Recommendation:** Implement more robust error handling, possibly using `Result` types instead of exceptions. Log specific errors rather than catching all exceptions.

### 4. FIXME and TODO Comments

The code contains several FIXME and TODO comments that indicate unresolved issues:

```ocaml
(* FIXME attention, le basename détruit les sous dossiers *)
(* TODO handle rn errors *)
```

**Recommendation:** Address these FIXMEs and TODOs before finalizing the PR. The basename issue particularly could lead to data loss when subdirectories are involved.

### 5. Debug Print Statements

There are several debug `Printf.eprintf` statements that should be removed:

```ocaml
let get_portrait_name_aux conf base p saved =
  let name = default_image_filename_aux "portraits" base p saved in
  let f = Filename.concat (portrait_folder conf) name in
  Printf.eprintf "Get_portrait_name: %s, %s\n" name f;  (* Debug print *)
  match find_img_opt f with
  | Some (`Path p) ->
      Printf.eprintf "Path %s/n" p;  (* Debug print with typo: /n not \n *)
      Filename.basename p
  | Some (`Url u) ->
      Printf.eprintf "Url %s/n" u;  (* Debug print with typo *)
      Filename.basename name ^ ".url"
  | None -> ""
```

**Recommendation:** Remove debug print statements or replace them with proper logging.

### 6. Interface Changes

The function `path_of_filename` has changed its return type from `` `Path of string`` to just `string`, which could cause bugs at call sites:

```ocaml
(* Before *)
let path_of_filename conf fname =
  let fname1 = Filename.concat (!GWPARAM.images_d conf.bname) fname in
  if Sys.file_exists fname1 then `Path fname1
  else `Path (Util.search_in_assets (Filename.concat "images" fname))

(* After *)
let path_of_filename conf fname =
  let fname1 = Filename.concat (!GWPARAM.images_d conf.bname) fname in
  if Sys.file_exists fname1 then fname1
  else Util.search_in_assets (Filename.concat "images" fname)
```

**Recommendation:** Verify all call sites are updated to handle the new return type. Consider keeping the function interface consistent.

### 7. UI Templates Complexity

The UI templates, especially `carrousel.txt`, are complex with many conditional sections:

```html
%if;(has_blason or has_blason_stop or has_old_blason)
<h1 class="display-4 mt-2">[*blason/blasons]0</h1>
%if;has_blason_self;
  <div class="d-flex align-self-center mt-1">
    <!-- Complex nested conditionals -->
  </div>
%elseif;has_blason_stop;
  <!-- Another conditional section -->
%elseif;has_blason;
  <!-- Yet another conditional section -->
%end;
```

**Recommendation:** Consider simplifying templates by extracting common patterns into macros or separate template files.

### 8. Duplicate Code

There's a fair amount of duplicate code, particularly for handling portraits and blasons:

```ocaml
let get_portrait_aux conf base p saved = ...
let get_portrait conf base p = get_portrait_aux conf base p false
let get_old_portrait conf base p = get_portrait_aux conf base p true

let get_blason_aux conf base p self saved = ...
let get_blason conf base p self = get_blason_aux conf base p self false
let get_old_blason conf base p self = get_blason_aux conf base p self true
```

**Recommendation:** Consider refactoring to reduce duplication, possibly with a more generic image handling function that is parameterized by image type.

### 9. SQL Injection Risks in Blason URL Generation

The `string_of_blason_url` function constructs URLs using string concatenation from user data:

```ocaml
let k = Filename.basename fname |> Filename.chop_extension in
(* k is first_name.occ.surname.blason *)
let parts = String.split_on_char '.' k in
let access =
  Format.sprintf "p=%s&n=%s&oc=%s" (List.nth parts 0) (List.nth parts 2)
    (List.nth parts 1)
in
```

**Recommendation:** Use proper URL encoding for all parameters to prevent potential injection issues.

### 10. Unsafe Array Access

The code includes some array accesses without bounds checking:

```ocaml
let parts = String.split_on_char '.' k in
Format.sprintf "p=%s&n=%s&oc=%s" (List.nth parts 0) (List.nth parts 2) (List.nth parts 1)
```

If `parts` doesn't have at least 3 elements, this will raise an exception.

**Recommendation:** Add bounds checking or use pattern matching to safely access array elements.

## Conclusion

The implementation is generally sound but would benefit from addressing the issues noted above. The most critical issues are:

1. The interface change to `path_of_filename` which could cause bugs at call sites
2. The file operation error handling which needs improvement
3. The basename issue with subdirectories which could lead to data loss
4. The debug print statements which should be removed

After addressing these issues, the code would be much more robust and ready for merging.

a2line committed:
Improve robustness of file saving mecanism

Enhances the move_file_to_save function to make it more reliable when handling files. The improvements include:

- Better error handling with specific error messages
- Proper cleanup of existing files in the save directory before moving
- More consistent handling of associated files (.txt and .src)
- Prevention of potential file loss when moving between directories

These changes address occasional file loss issues during the portrait/blason management operations and provide a more stable user experience.

a2line committed:
Template enhancements

a2line committed:
Modernize and improve carrousel JavaScript

efactored the JavaScript to improve maintainability and
follow modern jQuery best practices. The changes include:

- Replaced duplicate code with reusable generic functions
- Fixed minor bugs and typos in event handlers
- Used proper jQuery methods for DOM manipulation (prop vs attr)
- Organized code initialization with jQuery ready pattern
- Applied ES6 syntax for cleaner, more concise code

This refactoring maintains the same functionality while making the code
more robust and easier to maintain.

hgouraud commented:
le bouton "move blason to ancestor" ne fonctionne que si la personne porte le blason elle même
pas si elle en a hérité d'un ancêtre
il faut encadrer les lignes 265-281 avec un %if;has_blason_self; ... %end;

hgouraud committed:
move blason up only if has_blason_self

hgouraud committed:
ask for image name in URL case only

hgouraud committed:
copy_to_blason from image store. correct status report

a2line committed:
Replace refresh system with HTTP redirects, make image name optionnal

Remove complex REFRESH system requiring two HTTP requests:
- Implement direct HTTP 302 redirects for seamless UX
- Simplify template conditions (remove digest dependencies)
- Remove all digest validation for cleaner URLs and code

Make image name optional when importing from URL:
- Add extract_filename_from_url() to intelligently parse filenames from URLs
- Automatically extract clean filename (remove protocol, URL parameters)
- Update JavaScript to show extracted filename as placeholder
- Fallback to timestamp-based name for invalid URLs
- User can still override the extracted name if desired
- Clean JavaScript for image name field management

Bug fixes:
- Fix .url file deletion when filename contains multiple dots (image.png.url)

a2line committed:
fix carrousel directory renaming

* remove useless and confusing `loop` function
* calculate new/old keys with good normalisation (lowercases + space to underscores)
* rename the carrousel directory if it exists
* rename all portrait/blason files iterating their associated extensions
* handle errors via Logs.syslog without breaking the process

a2line committed:
fix links to individuals with apostrophe

as suggested by mdelambilly
